packages/@sanity/cli/src/actions/login/login.ts

```diff
@@ -106,77 +130,91 @@ export async function login(
   output.print(chalk.green('Login successful'))
 }
 
-function getAuthChannel(baseUrl: string, provider: LoginProvider, iv: string): EventSource {
-  const uuid = crypto.randomBytes(16).toString('hex')
-  let listenUrl
-  if (provider.type === 'saml') {
-    listenUrl = `${baseUrl}/v2021-10-01/auth/saml/listen/${provider.id}/${uuid}?iv=${iv}`
-  } else {
-    listenUrl = `${baseUrl}/v1/auth/listen/${provider.name}/${uuid}?iv=${iv}`
-  }
-  return new EventSource(listenUrl)
-}
+function startServerForTokenCallback(options: {
+  apiHost: string
+  apiClient: CliApiClient
+}): Promise<{token: Promise<TokenDetails>; server: Server}> {
+  const {apiHost, apiClient} = options
+  const domain = apiHost.includes('.sanity.work') ? 'www.sanity.work' : 'www.sanity.io'
+
+  const attemptPorts = callbackPorts.slice()
+  let callbackPort = attemptPorts.shift()
+
+  let resolveToken: (resolvedToken: TokenDetails | PromiseLike<TokenDetails>) => void
+  let rejectToken: (reason: Error) => void
+  const tokenPromise = new Promise<TokenDetails>((resolve, reject) => {
+    resolveToken = resolve
+    rejectToken = reject
+  })
 
-function getAuthInfo(es: EventSource): Promise<{secret: string; url: string}> {
-  const values = {secret: '', url: ''}
   return new Promise((resolve, reject) => {
-    es.addEventListener('error', (err: MessageEvent | EventWithMessage) => {
-      es.close()
-      const message = 'message' in err && typeof err.message === 'string' ? err.message : ''
-      reject(new Error(`Unable to get authorization info: ${message}`))
-    })
+    const server = http.createServer(async function onCallbackServerRequest(req, res) {
+      function failLoginRequest() {
+        res.writeHead(303, 'See Other', {Location: `https://${domain}/login/error`})
```

Comment:
If there are any error messages we would want to display to the user, the error page accepts various error codes. (i.e `/login/error?error=NO_SESSION_ID`)